Write module-level `qiskit.circuit` documentation

jakelishman opened the pull request:
### Summary

I was pretty surprised to see that we had very little written for our principal module, and the first entry in our API docs.

This is a totally new set of documentation for the `qiskit.circuit` module, including discussion of what circuits are, what they represent in Qiskit, how all the different data elements fit together, and where to go for more information on things.

This commit is intended to be followed by a later one that does a similar rewrite for the whole `QuantumCircuit` class, including structuring its references to all its attributes and methods, and adding a holistic view of how the methods fit together.  This is why there are some (hidden to Sphinx) `TODO` comments scattered throughout this documentation.

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Details and comments

I consider this PR to be about half the work I was intending to do; the other is to do a similar thing to this to the `QuantumCircuit` page itself.  Some notable concepts I was intending to include in that were:

- actual grouping structure to how the methods are presented
- `append`, `_append` and `compose` and their differences
- the control-flow builder interfaces
- what `layout` is and when it's populated

I tweaked some of the documented classes here to remove redundant or useless information, but I didn't attempt any complete rewrites of every single object in `qiskit.circuit`, so there's likely still many places that could benefit from short examples.  I also think there's places where we might want to link to "tutorial-like" documentation elsewhere on docs.quantum.ibm.com; this is API-level documentation and primarily concerns how to interact with the objects at a fairly local level, and how their APIs fit together to build up the API of the whole `circuit` module.

As a minor code change, this formally documents the `SessionEquivalenceLibrary` module attribute and makes it available from the `qiskit.circuit` root.

---

[Here's a link to the Azure run for the first commit](https://dev.azure.com/qiskit-ci/qiskit/_build/results?buildId=55004&view=results), from which you can hopefully find the docs build as the "1 artifact" under the "Preliminary tests" stage. Admittedly it just uses Sphinx's built-in Alabaster theme, so isn't really representative of what it would look like on IQP - it's just what would happened with "unthemed" Sphinx.  (Weirdly, the docs artefact seems to overlap the left sidebar and the content for me, which it didn't do from a local build.)

jakelishman committed:
Write module-level `qiskit.circuit` documentation

I was pretty surprised to see that we had very little written for our
principal module, and the first entry in our API docs.

This is a totally new set of documentation for the `qiskit.circuit`
module, including discussion of what circuits are, what they represent
in Qiskit, how all the different data elements fit together, and where
to go for more information on things.

This commit is intended to be followed by a later one that does a
similar rewrite for the whole `QuantumCircuit` class, including
structuring its references to all its attributes and methods, and adding
a holistic view of how the methods fit together.  This is why there are
some (hidden to Sphinx) `TODO` comments scattered throughout this
documentation.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -10,38 +10,62 @@
+on a predefined amount of quantum data (in Qiskit, we only directly support qubits) with unitary
+operations (gates), measurements and resets.  In addition, a quantum circuit can contain operations
+on classical data, including near-time computations and control-flow constructs, which are executed
+by the controllers of the QPU.
```
```suggestion
A fundamental model of quantum computations is the *quantum circuit*.  A quantum circuit represents a computational routine that can be run on a quantum processing unit (QPU).  A quantum circuit acts
on a predefined amount of quantum data with unitary operations (quantum gates), measurements and resets.  In addition, a quantum circuit can contain operations over classical and quantum data, including near-time computations and classical control flow constructs, which are executed by the classical control hardware 
of the QPU. A quantum circuit is typically executed multiple times on a QPU to gather statistics about the prepared quantum data. One such quantum circuit execution, including measurements, is called a *shot*. Qiskit natively supports quantum bits (qubits) as quantum data.
```
I think we should write one sentence about "near-term computations" here before using it...

jakelishman replied:
"executed by the QPU control hardware" kind of _is_ the definition of "near-time compute", but we can change this pending your other comment about the use of "near time" as a description.

jakelishman replied:
I think this point was largely resolved by the terminology switch to "real time", which is more immediately descriptive anyway.

sbrandhsn reviewed (commented):
I'm looking forward to reviewing this in detail. How do you feel about references to the qiskit textbook?

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -10,38 +10,62 @@
+
+    You may wish to skip the introductory material and jump directly to:
+
+    * :ref:`the API overview of the whole circuit module <circuit-module-api>`
```
```suggestion
    * :ref:`the API overview of the quantum circuit module <circuit-module-api>`
```

jakelishman replied:
I don't think we need to put the word quantum everywhere - it's clear from context after the first couple of goes, so after a while it just adds noise.

Eric-Arellano commented:
See https://github.com/Qiskit/documentation/pull/929 for a preview of this in IQP and instructions on how to run it locally and also regenerate the docs.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -10,38 +10,62 @@
+    * :ref:`the API overview of the whole circuit module <circuit-module-api>`
+    * :ref:`the detailed discussion about how circuits are represented <circuit-repr>`
+    * the core :class:`QuantumCircuit` class for how to build and query circuits
+    * :ref:`information on construction custom instructions <circuit-custom-gates>`
```
```suggestion
    * :ref:`information on the construction of custom operations <circuit-custom-gates>`
```
We haven't talked about instructions, yet. Should we rephrase quantum/classical operations as instructions?

jakelishman replied:
This link is specifically about the `Instruction` class, not `Operation`. The reason I didn't use the class name as a Sphinx cross-ref directly is just because it's really tricky/annoying to nest inline syntax elements in rST (unfortunately).

jakelishman replied:
Also, about "not defined yet", that's deliberate - this is a quick set of links for people already familiar with Qiskit and all the introductory content, like people who are coming back to the page for a second time. It's like a table of contents - I think it's ok to use this terminology in it.

sbrandhsn replied:
Alright, makes sense to me. :-)

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -10,38 +10,62 @@
+
+Circuits are the low-level construct that are used to build up the higher level :ref:`primitives of
+quantum computation <qiskit-primitives>`, which accumulate many shots of quantum-circuit execution
+along with advanced error-mitigation techniques and measurement optimizations.
```
```suggestion
In the Qiskit software project, quantum circuits are low-level constructs that are used to build up objects that are on a higher level of abstraction such as :ref:`primitives of
quantum computation <qiskit-primitives>`. Primitives accumulate encapsulate quantum circuit executions with advanced error-mitigation techniques and measurement optimizations.
```

jakelishman replied:
I'm not sure your text is quite correct about the minimum requirements of a primitive. The absolute requirement of a primitive is some sort of defined accumulation of several shots into a collection of classical data - all the error mitigation stuff is extras that can be applied but needn't (for example the reference ones don't do any).

Which bit were you aiming to clarify here?

sbrandhsn replied:
I think I wanted to clarify:
- The meaning of 'level': it could mean closer to the hardware, more/less abstract, hierarchical or even good/bad.
- If you don't already know about error-mitigation and measurement optimization you might wonder what "accumulate shots" has to do with the former.
I don't think addressing these is absolutely necessary but I think it would make this bit more informative.

sbrandhsn replied:
Also, I thought it is important to clarify that now we are speaking about circuits as a software object instead of a concept/model of quantum computation...

jakelishman replied:
> clarify that now we are speaking about circuits as a software object

Actually, I'd argue that at this point in the text, we're not entirely making that distinction - we do that in the two-to-three paragraphs after this.  I'd argue that this paragraph is still consistent with (somewhat fuzzy) definition of a "circuit" from outside the context of software.

> If you don't already know about error-mitigation and measurement optimization you might wonder what "accumulate shots" has to do with the former.

I'll tweak the text a little to make this clearer, thanks.  I don't think we necessarily need to define all the error-mitigation terms or anything, because they're not directly relevant to the discussion of `qiskit.circuit`; they'd be more appropriate for external learning material and/or the Primitives API documentation, if that needs to refer to them.

jakelishman replied:
I've tweaked the terminology around the primitives bit to make it clear that "levels" referred to abstractions, and to expand the detail around how the primitives accumulate bitwise data into well-typed classical data with error statistics in 2c1cc17232a - I think that hopefully addresses your concerns?

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -10,38 +10,62 @@
+
+* a *physical* circuit, which is defined in terms of the *hardware qubits* of one particular
+  backend, and contains only operations that this backend natively supports.  You might also see
+  this concept referred to as an *ISA circuit*.
```
```suggestion
In Qiskit, quantum circuits are often distinguished by their compatibility with the QPU:

* an *abstract* quantum circuit, which is defined in terms of *virtual qubits* and arbitrary high-level
  operations, like encapsulated algorithms and user-defined gates.

* a *physical* quantum circuit, which is defined in terms of the *hardware qubits* of one particular
  backend, and contains only operations that this backend natively supports.  You might also see
  this concept referred to as an *ISA circuit*, i.e. a quantum circuit hat is compatible to a certain QPU.
```

jakelishman replied:
I don't think we need the additional text on the end here - it's just repeating the previous sentence. I'm worried that the change in the first sentence also weakens the distinction, which is pretty vital for us to make clear with the changes to submission that are coming in Runtime.

sbrandhsn replied:
I tend to disagree that it is only repeating the former sentence. This is the third paragraph of the text and there are two terms that users may not be familiar with ("hardware qubits" and "ISA"). I therefore thought it makes sense to add one sentence for uninitiated users.

I see your second point but was wondering what "regime" might be in "can be defined in one of two regimes". However it can be inferred by the enumeration so this is just a minor point in my opinion...

jakelishman replied:
I guess I'd rather leave additional definitions for the full definitions section below.  Perhaps an alternative is to move this little lead-in to below the definitions, if you think it's important that everything is exactly defined before use?

sbrandhsn replied:
I think it is fine. :-)

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -10,38 +10,62 @@
 
-In Qiskit, this core element is represented by the :class:`QuantumCircuit` class.
-Below is an example of a quantum circuit that makes a three-qubit GHZ state
+In Qiskit, this core element is represented by the :class:`QuantumCircuit` class.  Below is an
```
```suggestion
In Qiskit, a quantum circuit is represented by the :class:`QuantumCircuit` class.  Below is an
```

jakelishman replied:
Done in 2c1cc17232a.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+================================
+
+There is a lot of specialized terminology around quantum circuits.  Much of this is common in
+quantum-computing literature, while some is more specific to quantum software packages, and a small
+amount specific to Qiskit.  This is an alphabetical listing of some of the important concepts as a
```
```suggestion
quantum computing literature, while some is more specific to quantum software packages such as Qiskit.  This is an alphabetical listing of some of the important concepts as a
```

jakelishman replied:
I can make this clearer if you like, but your suggestion changes the intended meaning. The idea is that some of the terminology is totally general, some is jargon related to all quantum software, some is jargon related specifically to Qiskit - that's why the last two elements were separated.

(It's also neither here nor there, but the removed hyphen wasn't wrong - "quantum-computing" here is a compound adjective modifying the noun "literature", and compound adjectives are typically hyphenated.)

sbrandhsn replied:
Sure, I understood that - my thinking was that qiskit, as one of the largest software packages, would define the terminology specific to software packages anyway so that distinction is not absolutely necessary but I get your point. :-) 

I will need to look a bit more into compound adjectives, thanks. :-) It just looked incorrect when I went over it first and google appears to confirm that it is rather unusual to write it with a hyphen (6 thousand results vs 77 million)...

jakelishman replied:
We're big, but I wouldn't say we're an entire monopoly, and it's probably (imo) best to recognise that others might say things a bit differently to us.

> google appears to confirm that it is rather unusual to write it with a hyphen

tbf, that's because it usually appears with "computing" as the noun, in which case the hyphen would be unambiguously wrong, and because not everyone uses the hyphen in compound adjectives anyway.  It's still valid to use one (in both British and American English).

sbrandhsn replied:
Sounds good to me.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+hardware qubit
+    The representation of a single qubit on a particular *QPU*.  A hardware qubit has some physical
+    quantum-mechanical system backing it; unlike a *virtual qubit*, it has particular coupling
+    constraints and only certain gates can be applied to certain physical qubits.
```
```suggestion
    The representation of a single qubit on a particular *QPU*.  A hardware qubit is represented by a physical
    quantum mechanical system such as superconducting circuits; unlike a *virtual qubit*, it has particular 
    constraints on permissible interactions, i.e. only certain quantum gates can be applied to certain sets of 
    physical qubits in a QPU.
```

jakelishman replied:
Done (partially) in 2c1cc17232a.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+    particular *QPU*.  For example, one QPU may allow :math:`\sqrt X` and :math:`R_Z` operations on
+    all single hardware qubits, and :math:`CX` operations on certain pairs of hardware qubits.
+
+logical qubit
+    A collection of several *physical qubits* that are controlled together by a QPU (from the user's
+    perspective) to apply real-time quantum error correction.  A logical qubit is a type of
+    *hardware qubit* for Qiskit.
```
I think we should not talk about logical qubits at all.

jakelishman replied:
Why? I think it's very important to tell people the level of abstraction that Qiskit treats them at, because otherwise it could seem like Qiskit won't be performing any abstraction over them.

sbrandhsn replied:
I think the discussion on logical qubits can be left out here because Qiskit does not really process data on that level of abstraction, right? I also think that the concept of logical qubits can be quite orthogonal to the concept of physical and virtual qubits introduced here. For me, logical qubits (or: error-corrected quantum circuits) are defined by their encoding/decoding/gate operation protocol. For these aspects, the qubits can be physical or virtual.

I'm worried that users that know about error-correction will wrongfully assume error correction can be processed in Qiskit natively or there exist abstractions between logical, physical and virtual qubits when in fact, Qiskit does not care about what the qubit do or how they are realized in the end... Maybe the discussion around logical qubits should be reduced to this statement. :-)

User that do not know about error correction will likely just learn some details that they won't be able to apply or control when using Qiskit.

jakelishman replied:
> I think the discussion on logical qubits can be left out here because Qiskit does not really process data on that level of abstraction, right? I also think that the concept of logical qubits can be quite orthogonal to the concept of physical and virtual qubits introduced here. For me, logical qubits (or: error-corrected quantum circuits) are defined by their encoding/decoding/gate operation protocol. For these aspects, the qubits can be physical or virtual.

Right, this is the absolute crux of why I think they _should_ be mentioned.  Your paragraph says effectively "you could do either", but in the model I'm describing here, it doesn't make sense to say that a logical qubit a virtual; a virtual qubit is something that exists in Qiskit _entirely_ separated from hardware - they only exist in abstract circuits.

The text you highlight here is largely a formalism of this statement:

> Qiskit does not care about what the [logical] qubit do or how they are realized in the end

and I think it's important to make that clear.  I also mentioned in the surrounding text that hardware vendors might choose to expose both a logical- and a physical-qubit target in the intermediate periods, so Qiskit _could_ be used for manual implementation of EC protocols - both fit within our abstraction of "hardware qubits".

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+    quantum-mechanical system backing it; unlike a *virtual qubit*, it has particular coupling
+    constraints and only certain gates can be applied to certain physical qubits.
+
+    Qiskit does not distinguish *logical qubits* from any individual *physical qubits* when talking
+    about hardware qubits.  A QPU may implement its hardware qubits as logical qubits, where each
+    hardware qubit comprises many physical qubits that are controlled and error-corrected opaquely
+    to Qiskit by the control electronics.  More likely, for near-term applications, a QPU will be
+    directly exposing its physical qubits as the hardware qubits for Qiskit to reason about.
+
+    Both physical and logical qubits will have coupling constraints between them, only permit
+    certain quantum operations on them, and have scheduling concerns between them.  Qiskit abstracts
+    these concerns together in the concept of hardware qubits. In the early days of quantum error
+    correction, particular backends may let you access their qubit resources either as high-level
+    logical qubits or as low-level physical qubits through Qiskit.
```
I think we should not talk about logical qubits at all.

jakelishman replied:
(I'll just close this one in favour of https://github.com/Qiskit/qiskit/pull/11904#discussion_r1506721949 to avoid duplication, since we'll need to go through more than just these two parts pending the resolution of that anyway.)

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+    amount of near-time classical computation available with particular *QPU*\ s will vary
+    significantly dependent on many factors, such as the controlling electronics and the qubit
+    technology in use. You should consult your hardware vendor's documentation for more information
+    on this.
```
I suggest to replace "near-term" with "at-runtime". I was confused when you used near-time compute in the introduction at first - here it becomes clear. :-)

jakelishman replied:
In theory, that's kind of the point of this definition section. That said, if you think "runtime" is going to be clear and not easily mixed up with "runtime of the Python program" or stuff like that, I can look at changing it everywhere?

sbrandhsn replied:
I'm not sure... The terminology offline/online/at-runtime also occurs in computer architecture but I agree that at-runtime might conflict with the python runtime and also with qiskit runtime. What do you think about "mid-circuit" or "in-circuit". What does 'near' refer to in 'near-time'? :-)

jakelishman replied:
Let's bring it up in the open qiskit-dev meeting and ask what terminology we want for this.

jakelishman replied:
From the dev meeting: the terminology from elsewhere in IBM Quantum is "real time", we think - I hadn't even considered that, but now I hear it I think it's the best choice.  Changed over in dd78e25a6c8.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+physical circuit
+    A *circuit* defined in terms of *hardware qubits* and only the quantum operations available in a
+    particular *QPU's* *ISA*.  Physical circuits are tied to one particular QPU architecture, and
+    will not run on incompatible architectures.  You may also hear this referred to as an *ISA
```
```suggestion
    will not run on other, incompatible architectures.  You may also hear this referred to as an *ISA
```

jakelishman replied:
Done in 2c1cc17232a.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+    TODO: actually write the "in-depth section on building circuits and cross-ref to it.
+
+The principal class is :class:`.QuantumCircuit`, which has its own documentation page, including
+an in-depth section on building circuits. The simplest quantum and classical data is built out of
```
```suggestion
an in-depth section on building circuits. The simplest representation of quantum and classical data is built out of
```

jakelishman replied:
I reworded this a bit more completely in 2c1cc17232a.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+"bits" and "registers":
+
+* :class:`Bit`, an atom of data
+    * :class:`Qubit`
+    * :class:`Clbit`
+    * :class:`AncillaQubit`
```
```suggestion
    * :class:`Qubit` - a quantum bit of quantum data
    * :class:`Clbit` - a classical bit (0 or 1, False or True) of classical data
    * :class:`AncillaQubit` - a quantum bit that provides 'scratch' space, e.g. for error correction or for more efficient realizations of gates
```

jakelishman replied:
I'm not clear this adds anything - we already defined the terms. We might want to add "ancilla" to the definitions, but the bits about qubit and clbit are repeating basic information we already gave.

sbrandhsn replied:
True, I was thinking about suggesting to remove AncillaQubits and AncillaRegister all together :-)

jakelishman replied:
Yeah, for sure we don't actually make any use of them right now.  We'd have to have a separate discussion for that, though, and they wouldn't be eligible for removal until Qiskit 2.0, and we'd still need to document them in the mean time.

jakelishman replied:
I added a brief definition of "ancilla qubit" to the definitions list in 2c1cc17232a, which hopefully is enough for this?

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+    * :class:`AncillaRegister`
+
+Within a circuit, each complete :class:`CircuitInstruction` is made up of an :class:`Operation`
+(which might be an :class:`Instruction`, a :class:`Gate`, or some other subclass) and the qubit
+and clbit operands.  The core base classes here are:
```
```suggestion
(which might be an :class:`Instruction`, a :class:`Gate`, or some other subclass) and the quantum
and classical bit operands.  The core base classes here are:
```

jakelishman replied:
These are standard Qiskit terms, and they're used everywhere in all our existing documentation too (and property names, method names, stuff like that). I can put in a definition of "clbit" to the top list if you like, but I think replacing all uses of "qubit" and "clbit" with "quantum bit" and "classical bit" will make the documentation more noisy rather than clearer.

sbrandhsn replied:
I think it makes sense to once define "classical bit (clbit)" somewhere in the text.

jakelishman replied:
Added a definition to the list in 2c1cc17232a.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+Some instructions are particularly special in that they affect the control flow or data flow of the
+circuit.  The top-level ones are:
+
+* :class:`Barrier`, to forbid optimizations from crossing a barrier
```
```suggestion
* :class:`Barrier`, defines parts of a the circuit that should be optimized by the transpiler separately, i.e. optimizations spanning over this barrier are not allowed
```

jakelishman replied:
Done (mostly) in 2c1cc17232a - I kept it to a single clause up here since we define it more completely in the actual API documentation around it.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+
+Programmatically, this class is actually implemented in Rust and is a constructed handle to internal
+data within Rust space.  Mutations to instances of this class will not be reflected in the circuit.
+In general, you cannot mutate instruction contexts that are already in the circuit directly; the
```
What about
`qc.data = [instr for instr in qc.data if instr.operation.num_qubits > 2]` as suggested in https://www.youtube.com/watch?v=cC0dFDv7XMQ ?
If we don't want user to do things like these we should probably not advertise them on youtube... :-)

jakelishman replied:
That one isn't a mutation of a `CircuitInstruction` instance, it's modification of the container of instructions.

The Qiskit team don't have control over the youtube channel, but I wouldn't have suggested that pattern had I been consulted - better to `copy_empty_like` and then call `append` in a loop (or `_append`).  (Fwiw, I think the `QuantumCircuit.data` setter actually will cope entirely correctly with this, but my problem is that I don't think `QuantumCircuit.data` should be exposed as mutable at all! Unfortunately for me, that's not the actual state, though.)

sbrandhsn replied:
Alright, makes sense to me. :-)

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+Operations, instructions and gates
+----------------------------------
+
+Within a :class:`CircuitInstruction`, the minimal interface that any operation must fulfil is
```
"fulfil" looks strange but apparently this is british english :-)

jakelishman replied:
Oh yeah, if you spot any Britishisms feel free to offer the Americanisations as suggestions - I _try_ and write with American spelling for these things, but I'm fighting 30 years of conditioning haha.

jakelishman replied:
Fixed in 2c1cc17232a.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+:ref:`circuit-custom-gates` for more detail.
+
+The :attr:`~Instruction.params` of an instruction can technically be arbitrary, but in general you
+should attempt to stick to parametrizations in terms of angles, wherever possible.  Qiskit itself
```
Can we generalize 'angles' to reals?

jakelishman replied:
We probably can. Angles were what these were originally meant to be (with an intent that control hardware might want to treat them with fixed-point integer semantics), but in practice, real numbers are probably fine.

sbrandhsn replied:
Sounds good! :)

jakelishman replied:
Done in 2c1cc17232a.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+not valid in floating-point arithmetic, and :class:`.expr.Expr` will not do this.
+
+The "compile-time" part of these parameters means that you typically will want to "assign" values to
+the parameters before sending them for execution.  These parameters can typically be used anywhere
```
```suggestion
the parameters before sending the circuit including them for execution.  These parameters can typically be used anywhere
```

jakelishman replied:
Done in 2c1cc17232a.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+    :no-inherited-members:
+
+These control-flow operations (:class:`IfElseOp`, :class:`WhileLoopOp`,
+:class:`SwitchCaseOp` and :class:`ForLoopOp`) all have specific state that defines the branching
```
```suggestion
:class:`SwitchCaseOp` and :class:`ForLoopOp`) all have a specific state that defines the branching
```

jakelishman replied:
I'm using "state" in the sense of "class state" - classes "have state" in the sense of "are stateful" rather than "have a state".

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+blocks manually will need to track these captures when building control-flow circuit blocks and add
+them to the block using :meth:`~QuantumCircuit.add_capture` (or the ``captures`` constructor
+argument), but user code will typically use the control-flow builder interface, which handles this
+automatically.
```
This is quite a lot of detail but I guess this is also the right location to put it into... :-)

jakelishman replied:
tbh I consider this amount of detail being necessary a strong indication that we didn't do a great job with the data model :(

sbrandhsn replied:
Do you think this should be part of the documentation on the quantum circuit?

jakelishman replied:
I put it here because this `qiskit.circuit` module is where the control-flow operations are documented and the module-level documentation is the place for all the stuff about how to manually fit together different parts of the `cirucit` API.  I will mention this in the discussion on the `QuantumCircuit` page as well for sure (mostly in the discussion of the `Var`-related methods).

sbrandhsn replied:
Sounds good.

sbrandhsn commented on qiskit/circuit/__init__.py:
```diff
@@ -54,81 +78,825 @@
+be safe to subclass, and have hook points for subclasses to implement.  If your custom gate is
+parameterless and stateless, you may also want to derive from the corresponding singleton class in
+:mod:`qiskit.circuit.singleton`, such as :class:`SingletonGate`.  You should consult the
+documentation in :mod:`qiskit.circuit.singleton` for additional methods and hookpoints for the
```
```suggestion
documentation in :mod:`qiskit.circuit.singleton` for additional methods and hook points for the
```

jakelishman replied:
Done in 2c1cc17232a.

sbrandhsn commented on qiskit/circuit/parameter.py:
```diff
@@ -25,10 +25,13 @@
 
 class Parameter(ParameterExpression):
-    """Parameter Class for variable parameters.
+    """A compile-time symbolic parameter.
```
```suggestion
    """A symbolic parameter whose value must be determined at compile-time, i.e. before sending the associated quantum circuit to a quantum processing unit.
```
I thought a bit more context would be warranted here, if the user did not read the circuit module documentation first...

jakelishman replied:
Done (slightly differently) in 2c1cc17232a.

sbrandhsn commented:
Thanks for writing this up! I learned a lot while reading this! :-) I will go over your comments tomorrow and sorry for submitting individual comments instead of one batch of comments with one code review. :s

jakelishman committed:
Apply suggestions from Sebastian's review

Co-authored-by: Sebastian Brandhofer <148463728+sbrandhsn@users.noreply.github.com>

jakelishman committed:
Move larger classes to separate pages

This moves the larger class definitions (basically those that actually
have any public methods) to separate pages, at the request of the docs
team.  Much of this is intended to be temporary, and for most of these
classes to be moved back inline once the IBM docs platform can better
support that.

jakelishman commented:
At the request of the docs team, I've temporarily moved most of the inline class documentation into `autosummary` tables in a564e59acd8.  Once the IBM docs platform can support them a little better, we can evaluate whether it's better to have them inline (imo, for most classes it's a "yes", but that's open for discussion).

jakelishman commented:
Sebastian: I've responded to all your comments, either with changes or follow-ups.  I marked quite a lot of conversations as "resolved" for the purposes of making open discussions easier to scan, but feel free to unresolve any you think I didn't adequately address.

jakelishman committed:
Fix docs build with autosummary

sbrandhsn commented:
Thanks @jakelishman! Let's discuss the remaining bits in the qiskit-dev meeting and then I think the circuit documentation is good to go. Thanks for writing this up! :-)

jakelishman committed:
Switch "near time"/"runtime" terminology to "real time"

This avoids overloading "runtime" with too many meanings, and appears to
be more consistent with other messaging coming out of IBM Quantum.

jakelishman commented:
Following discussion in the dev meeting today, I've switched all "near time" and "runtime" terminology to "real time" in dd78e25a6c8.

blakejohnson commented on qiskit/circuit/__init__.py:
```diff
@@ -10,38 +10,64 @@
-Overview
-========
+The fundamental element of quantum computing is the *quantum circuit*.  This is a computational
+routine that can be run, one shot at a time, on a quantum processing unit (QPU).  A circuit will act
```
What does "one shot at a time" add to the description? Are you trying to clarify that circuits are not expected to have shot loops?

jakelishman replied:
Part of what defines a circuit is, and more crucially, what it _isn't_, is that all the classical data from one circuit can be acquired in one run on a QPU.  I'm specifically trying to exclude the concept of a circuit containing an "expectation value" operation, because

- that's where we split the abstraction level and move to the primitives, and the split of abstraction layers is something we need to keep the software architecture understandable and maintainable
- it would really clash with how all the rest of circuit execution works in the model of being executed on hardware.

The idea being that an implementer of the primitives really wants some software concept of "this is a single shot", and the primitive implementation is what accumulates that into higher-typed classical data, especially since that's going to come with measurement statistics.

It's the accumulation of many shots that I'm trying to completely rule out - the circuit is the abstraction level that defines a single shot.  We _expect_ it to be run many many times in a shot loop during a primitive execution, but if it starts to represent the shot loop itself, it breaks the abstraction layer down between primitives and shots, which will make it harder to architect and work.

blakejohnson replied:
Ok, that makes sense. Thanks.

blakejohnson commented on qiskit/circuit/__init__.py:
```diff
@@ -10,38 +10,64 @@
+
+Circuits are at a low level of abstraction when building up quantum programs.  They are the
+construct that is used to build up to higher levels of abstraction, such as the :ref:`primitives of
+quantum computation <qiskit-primitives>`, which accumulate bitwise data from many shots of
```
> accumulate bitwise data

I think the "bitwise" qualifier here is unnecessary. A circuit in the future may return data following not `bit[n]` types, including `int`s, `angle`s, `complex`, etc. The sentence still works if you drop "bitwise".

jakelishman replied:
I'm ok to remove "bitwise" here if you still prefer, just a small argument in its defence first:

A circuit may return data that's not a `bit` type, but that data always would have been built up by switching on data that _was_ - that's the point I'm trying to emphasise.  I didn't see (say) IPE as being contrary to "bitwise"; the `angle` is still built up in a single circuit execution as bit-by-bit operations, and error statistics or smoothing of all those single-shot measurements would be the domain of a primitive.  For a `complex`, say, a single circuit execution can only build that up off single-bit measurements (however many of them), or things like `if` statements based on those that set the value.

Ofc, I suppose you could make the same argument about the primitives as well, it's just the "level" that the bitwise stuff happens is even further remove.

blakejohnson replied:
I was thinking of `complex` arising as the outcome of using the future Qiskit equivalent of OpenPulse's `measure_iq` which produces a single complex value as opposed to a single bit.

blakejohnson replied:
And I definitely was thinking about IPE when suggesting `angle` as a possible return type.

jakelishman replied:
Yeah, the OpenPulse enhanced measurements are a good point, and you could say the result of an `extern` call that's stateful in whatever way as well could easily give you structured data within the circuit.

I'll delete the word "bitwise" - I can totally buy that argument.

jakelishman replied:
Done in 2f4d541774.

jakelishman committed:
Remove misleading 'bitwise' terminology

1ucian0 commented on qiskit/circuit/store.py:
```diff
@@ -82,6 +86,7 @@ def rvalue(self):
         return self.params[1]
 
     def c_if(self, classical, val):
+        """:meta hidden:"""
```
I didnt know about this sphinx feature, thanks!

jakelishman replied:
It's only used by `autodoc` and it doesn't work `autosummary` (unfortunately), at least with our templates, so the idea that it works is quite new here.  My hope is that we can use this sort of thing more, though.

1ucian0 commented on qiskit/circuit/reset.py:
```diff
@@ -18,10 +18,13 @@
 
 class Reset(SingletonInstruction):
-    """Qubit reset."""
+    r"""Incoherently reset a qubit to the :math:`\lvert0\rangle` state."""
```
I'm feeling the word _incoherently_ might add more confusion than not...

jakelishman replied:
It's standard quantum-computing terminology, but if you have a preferred word, I can change it.

1ucian0 commented on qiskit/circuit/measure.py:
```diff
@@ -22,7 +22,10 @@ class Measure(SingletonInstruction):
-        """Create new measurement instruction."""
+        """
+        Args:
+            label: optional string label for this instruction.
```
should `duration` and `unit` be added too? Or are them inherited?

jakelishman replied:
They're not inherited, but we also haven't really documented them in any of the other instruction constructors.  If possible, I'd potentially like to leave changing the documentation around the calibration/scheduling-related components for another PR, if that's ok?  I think there's _a lot_ more work to be done about that, which I largely haven't touched.

1ucian0 commented on qiskit/circuit/instructionset.py:
```diff
@@ -124,27 +132,6 @@ def c_if(self, classical: Clbit | ClassicalRegister | int, val: int) -> "Instruc
-               qc.y(1).c_if(0, 1)
-
-               qc.draw('mpl')
-
```
why removing the example?

jakelishman replied:
I _think_ I originally did it because I also wrote:

> This :class:`InstructionSet` is now little used in Qiskit.  It provides a very minimal set of methods to perform post-append mutations on instructions (which *will* be propagated to the circuit), but these are now discouraged and you should use the alternatives noted in those methods.

in the main prose, but I also don't really remember and it may have not been entirely intended.  I can add it back for now if you prefer?

1ucian0 commented on qiskit/circuit/instruction.py:
```diff
@@ -269,12 +269,17 @@ def soft_compare(self, other: "Instruction") -> bool:
 
     def _define(self):
-        """Populates self.definition with a decomposition of this gate."""
+        """Populate the cached :attr:`_definition` field of this :class:`Instruction`.
```
is there any advantage to refer with `:attr:` if it has no docstring?

jakelishman replied:
Not really, but I don't think it particularly hurts either.  At one point I think I was considering documenting `_definition` as well - I can do that if you'd rather?

1ucian0 reviewed (approved):
Thanks for comment on my comments. Let's merge it!
